### PR TITLE
G1 Linked List

### PR DESCRIPTION
**Pull request by xyzhelen**
Helen, Lana, Luke

All tests passing

### CODE REVIEW

File: linked_list/main.py [hunk 1]
```diff
@@ -62,34 +63,94 @@ class LinkedList(AbstractLinkedList):
...
+        self.start = None
+        self.end = None
+            
+        if elements:
+            for element in elements:
+                self.append(element)
 
     def __str__(self):
-        pass
+        return '[' + ', '.join(str(i) for i in self) + ']'
```
Comment (martinzugnoni) [inline linked_list/main.py]: More pythonic ;-):

```
return '[{}]'.format(', '.join([str(i) for i in self]))
```

Concatenating strings with `+` is not very used, `.format()` is usually preferred.

File: linked_list/main.py [hunk 2]
```diff
@@ -62,34 +63,94 @@ class LinkedList(AbstractLinkedList):
...
     def __iadd__(self, other):
-        pass
+        return self.__add__(other)
 
     def __eq__(self, other):
-        pass
+        if self.count() != other.count():
+            return False
+        
+        for x, y in zip(self, other):
```
Comment (martinzugnoni) [inline linked_list/main.py:105]: Lovely! ;-)

**Comment by martinzugnoni** [comment]
Code is looking pretty good guys! Good job! :)

File: linked_list/main.py [hunk 3]
```diff
@@ -42,54 +42,103 @@ class Node(object):
...
     """
-
     def __init__(self, elements=None):
-        pass
+        self.start = None
+        self.end = None
+            
+        if elements:
+            for element in elements:
+                self.append(element)
```
Comment (eliloz) [inline linked_list/main.py:69]: I really like this init, we ended up using something very similar.

Comment (pbhag) [inline linked_list/main.py:69]: Nice usage of the append() in the **init** , much more efficient than what we were trying to come up with initially.

File: linked_list/main.py [hunk 4]
```diff
@@ -42,54 +42,103 @@ class Node(object):
...
 
     def __str__(self):
-        pass
+        return '[{}]'.format(', '.join(str(i) for i in self))
 
     def __len__(self):
-        pass
+        return sum(1 for i in self)
```
Comment (eliloz) [inline linked_list/main.py:75]: This is much simpler than what we implemented. We basically kept a running tally that was updated with every append() or pop() and returned that instead. I'll have to keep this in mind for future reference

File: linked_list/main.py [hunk 5]
```diff
@@ -42,54 +42,103 @@ class Node(object):
...
+        current = self.start
+        while current:
+            yield current
+            current = current.next
 
     def __getitem__(self, index):
-        pass
+        for i, node in enumerate(self):
+            if i == index:
+                return node.elem
```
Comment (pbhag) [inline linked_list/main.py:87]: This is a really efficient way of doing the **getitem**. We ran through our list, starting from the start node, but the enumerate() function was really smart -you didnt have to traverse the list, you guys were able to just locate the indices. nice

File: linked_list/main.py [hunk 6]
```diff
@@ -42,54 +42,103 @@ class Node(object):
...
 
     def __add__(self, other):
-        pass
+        new_list = []
```
Comment (eliloz) [inline linked_list/main.py:89]: I thought other was passed in as a a list anyway, so I'm not sure if making a new list was necessary. I could be wrong though

**Comment by an00bhav** [comment]
Just reading out code of the other group members:

1) Excellent way to combine the `isinstance` and the eq check for __eq__ for Node (line 53) 

2) Interesting way to implement the len method. +1

3) enumerate for __getitem__ is interesting too. Never thought about it. 

4) __add__  Could have used deepcopy from copy standard library to get rid of one for loop

Great code, nice concepts, got a lot to learn. Thanks a lot and keep it up.

**Pull request closed without merging**